### ISSUE TITLE
aws_ecs_task_definition and continuous delivery to ecs

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#632]
_This issue was originally opened by @dennari as hashicorp/terraform#13005. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

With the task and container definition data sources I'm *almost* able to get our continuous delivery setup to play nicely with Terraform. We rebuild the docker image with a unique tag at every deployment. This means that after the CI service redeploys a service, the corresponding task definition's *revision* is incremented and the *image* field in a container definition changes.

I dont' seem to be able to create a setup where the task definition could be managed by Terraform in this scenario.

### Terraform Version
v0.9.1
### Affected Resource(s)
- resource aws_ecs_task_definition
### Terraform Configuration Files

```hcl
# Simply specify the family to find the latest ACTIVE revision in that family.
data "aws_ecs_task_definition" "mongo" {
  task_definition = "${aws_ecs_task_definition.mongo.family}"
}
data "aws_ecs_container_definition" "mongo" {
  task_definition = "${data.aws_ecs_task_definition.mongo.id}"
  container_name  = "mongodb"
}

resource "aws_ecs_cluster" "foo" {
  name = "foo"
}

resource "aws_ecs_task_definition" "mongo" {
  family = "mongodb"
  container_definitions = <<DEFINITION
[
  {
    "cpu": 128,
    "environment": [{
      "name": "SECRET",
      "value": "KEY"
    }],
    "essential": true,
    "image": "${aws_ecs_container_definition.mongo.image}",
    "memory": 128,
    "memoryReservation": 64,
    "name": "mongodb"
  }
]
DEFINITION
}

resource "aws_ecs_service" "mongo" {
  name          = "mongo"
  cluster       = "${aws_ecs_cluster.foo.id}"
  desired_count = 2
  # Track the latest ACTIVE revision
  task_definition = "${aws_ecs_task_definition.mongo.family}:${max("${aws_ecs_task_definition.mongo.revision}", "${data.aws_ecs_task_definition.mongo.revision}")}"
}
```

The problem is then that after a CI deployment, terraform would like to create a new task definition. The task definition *resource* here points to an earlier revision and the image field is considered changed.

With the deprecated template resources, I was able to ignore changes to variables which solved this issue. One solution that comes to mind would be the ability to set revision of the `aws_ecs_task_definition` resource.

I'd be grateful for any and all insights.

**Comment by kurtwheeler** [comment]
I have run into this issue as well. I think the solution I am going to go with is to not have the task definition be managed by terraform. Circle CI has a [blog post](https://circleci.com/docs/1.0/continuous-deployment-with-aws-ec2-container-service/) about how to push a new task definition via a [script they provide](https://github.com/circleci/go-ecs-ecr/blob/master/deploy.sh).

I agree that the ability to set the `aws_ecs_task_definition` would enable managing the task definition via Terraform. However philosophically it does seem to break Terraform's model of having `resource` blocks correspond to resources within AWS. If that parameter were added then a single `aws_ecs_task_definition` resource block would be responsible for creating multiple AWS resources.

**Comment by JDiPierro** [comment]
I've gotten around this by using `terraform taint` as part of the deploy process:

  - Push the container to the ECR
  - `terraform taint` the task-def
  - `terraform apply`

**Comment by naveenb29** [comment]
@JDiPierro  While using the `taint` solution - does it not kill the current task and replace rather than deploy the new task and then drain the old one ?

**Comment by JDiPierro** [comment]
@naveenb29 Nope. I believe that would be the case if you were tainting the ECS Service. Since just the task-def is being recreated the ECS service is updated causing the new tasks to deploy. ECS waits for them to become healthy and then kills the old containers.

**Comment by tomelliff** [comment]
Our CI process is tagging the image as it's pushed to ECR and then passing that tag to the task definition. This automatically leads to the task definition changing so that Terraform knows to recreate it and then that's linked to the ECS service causing it to update the service with the new task definition.

I think I'm missing the issue that others are having here.

That said, I'd like for Terraform to (optionally) wait for the deployment to complete: new task definition has running tasks equal to the desired count and, potentially, that the other task definitions have been deregistered so no new traffic will reach them. I can't see a nice way to get at that information using the API though as the events don't really expose enough information so you'd probably have to describe the old task def, find the running tasks using it, find the ports they were running on, check the ALB they are registered to has all of those ports set as draining.

For now I'm simply shelling out and waiting until the PRIMARY service deployment has a running count equal to the desired count (doesn't catch the short period from PRIMARY tasks being registered to old tasks being deregistered) or waiting until the deployment list has a length of 1 (all old task definitions have been completely drained which is overkill as new connections won't arrive there so the deployment can be considered completed before this).

**Comment by radeksimko** [comment]
Hi folks,
see more detailed explanation of why this is happening at https://github.com/terraform-providers/terraform-provider-aws/issues/13#issuecomment-318700677

Labelling appropriately...

**Comment by dmikalova** [comment]
I was able to solve the inactive task definition issue with the example in the [ECS task definition data source](https://www.terraform.io/docs/providers/aws/d/ecs_task_definition.html). You set up the ECS service resource to use the the max revision of either what your Terraform resource has created, or what is in the AWS console which the data source retrieves.

The one downside to this is if someone changes the task definition, Terraform will not realign that to what's defined in code.

**Comment by Esya** [comment]
What do you guys think about having a remote backend (It's S3 in my case), and having your CI pipeline create the new task definition and changing the **.tfstate** file directly to match it ?

For example, mine looks like this : 

```javascript
"aws_ecs_task_definition.backend_service": {
    "type": "aws_ecs_task_definition",
    "depends_on": [
        "data.template_file.task_definition"
    ],
    "primary": {
        "id": "backend-service",
        "attributes": {
            "arn": "arn:aws:ecs:eu-west-1:REDACTED:task-definition/backend-service:8", // This could be changed manually
            "container_definitions": "REDACTED",
            "family": "backend-service",
            "id": "backend-service",
            "network_mode": "",
            "placement_constraints.#": "0",
            "revision": "8", // This could be increased manually
            "task_role_arn": ""
        },
        "meta": {},
        "tainted": false
    },
    "deposed": [],
    "provider": ""
},
```

Couldn't we just change the **arn** and the **revision**, so that the next time terraform runs, it still thinks it has the "latest version" of the task defintion in it's state ?

**Comment by chriswhelix** [comment]
I'm not sure I understand the problem y'all are trying to solve here. Why not just use terraform to create the new task definition in the first place, and then your tf state is always consistent? Our setup is similar to what @tomelliff describes.

**Comment by Esya** [comment]
@chriswhelix Well in my particular case, I have two separate repositories. One that holds the terraform project, and I have multiple workspaces for separate environments, and this creates my ECS cluster, my loadbalancer, my services, and the initial task definition.

On the other hand, for the project I'm trying to deploy, I'd like to have some CI/continuous delivery flow in place (Using gitlab pipelines in my case) to *"containerize"* the project, push it to ECR, and trigger a service update on my ECS cluster. 

So, when you say *"use terraform to create the new task definition in the first place"*, are you implying that on our CI system, we should also clone our terraform repo, change the variable that holds the image tag or that service, do a terraform apply, and commit + push to the TF repository?

tl;dr: Need a way to trigger service updates from the particular projects' build pipeline, without any user interaction with terraform.

**Comment by chriswhelix** [comment]
@Esya what we do is that each project has in its build config the version of the terraform repo it is expecting to be deployed with. When the CI pipeline is ready to deploy, it pulls down the terraform repo using the git tag specified in the project build config, then runs terraform against that, providing the image tag it just wrote to the ECR repo as an input variable.

We don't write down the ECR image tag in the terraform repo; it must be provided each time terraform is run. So that avoids simple code updates to projects requiring any change to the terraform repo.

**Comment by schmod** [comment]
I'm using [ecs-deploy](https://github.com/fabfuel/ecs-deploy) in my deployment pipeline, and a terraform config that looks something like this:

```hcl

# Gets the CURRENT task definition from AWS, reflecting anything that's been deployed
# outside of Terraform (ie. CI builds).
data "aws_ecs_task_definition" "task" {
  task_definition = "${aws_ecs_task_definition.main.family}"
}

# "Dummy" application for initial deployment
data "aws_ecr_repository" "sample" {
  name = "sample"
}

# ECR Repo for our actual app
data "aws_ecr_repository" "main" {
  name = "${var.ecr_name}"
}

resource "aws_ecs_task_definition" "main" {
  family = "${var.name}"
  task_role_arn = "${module.iam_roles.ecs_service_deployment_role_arn}"

  container_definitions = <<DEFINITION
[
  {
    "name": "${var.name}",
    "image": "${data.aws_ecr_repository.sample.repository_url}:latest",
    "essential": true,
    "portMappings": [{
      "containerPort": ${var.container_port},
      "hostPort": 0
    }]
  }
]
DEFINITION
}

resource "aws_ecs_service" "main" {
  name = "${var.name}"
  cluster = "${var.cluster}"
  desired_count = 2
  task_definition = "${aws_ecs_task_definition.main.family}:${max("${aws_ecs_task_definition.main.revision}", "${data.aws_ecs_task_definition.task.revision}")}"
  iam_role = "${module.iam_roles.ecs_service_deployment_role_arn}"

}
```

During the initial deployment, Terraform deploys an "empty" container.  When the CI pipeline runs,  ecs-deploy creates a new task definition with the relevant image/tag, and updates the service accordingly.  

Terraform recognizes these new deployments via `data.aws_ecs_task_definition.task,` and doesn't attempt to overwrite them.  

HOWEVER, if the task definition changes, Terraform will redeploy the sample application.  Hypothetically, `data.aws_ecs_container_definition` could be used to pull the image of the currently-active task definition.  However, I haven't been able to figure out a way to use this that doesn't create a circular dependency or result in a chicken/egg problem during the initial deployment (ie. the data source is looking for a task definition that hasn't been created yet):

```hcl
data "aws_ecs_task_definition" "task" {
  task_definition = "${aws_ecs_task_definition.main.family}"
}

data "aws_ecs_container_definition" "task" {
  task_definition = "${data.aws_ecs_task_definition.task.id}"
  container_name  = "${var.name}"
}

resource "aws_ecs_task_definition" "main" {
  family = "${var.name}"
  task_role_arn = "${module.iam_roles.ecs_service_deployment_role_arn}"

  container_definitions = <<DEFINITION
[
  {
    "name": "${var.name}",
    "image": "${data.aws_ecs_container_definition.task.image}",
  }
]
DEFINITION
}
```

This creates a cycle, _and_ won't work during the initial deployment.  

This is _very_ close to my ideal setup.  If Terraform somehow supported a "get or create" datasource/resource hybrid, I'd be able to do almost exactly what I'm looking for.

**Comment by chriswhelix** [comment]
@schmod you could possibly use a var to create a special bootstrapping mode, i.e. "count = var.bootstrapping ? 0 : 1" to turn on/off the data sources, and coalesce(data.aws_ecs_container_definition.task.*.image, "fake_image") on the task def.

I feel like if you're going to manage a particular resource with terraform, it's really best to make all modifications to it using terraform, though. If you solve this issue for container images, you're just going to have it again for service scaling, and again for environment changes, and again for anything else ecs-deploy does behind terraform's back.

What we really need is good deployment tools that work with terraform instead of around it.

**Comment by schmod** [comment]
Given the scope of what Terraform is allowed to do to my AWS resources, I'm rather apprehensive about running it in an automated/unmonitored environment.  On the other hand, I can control exactly what ecs-deploy is going to do.  

Infrastructure deployments and application deployments are very different in my mind.  There's a fairly large and mature ecosystem around the latter, and I don't think that Terraform should need to reinvent that wheel.  It should merely provide a configuration interface to tell it the exact set of changes that I expect those external tools to make.  

We already have a version of that in the form of the `ignore_changes` lifecycle hook.  My problem could also be solved if we supported `container_definition` as a 1st-class citizen (similar to `aws_iam_policy_document`), and allow something like `ignore_changes=["container_definition.image"]`.

**Comment by chriswhelix** [comment]
@schmod isn't the real issue what your build agent is permissioned to do? If your build agent has least privileges for the changes you actually want it to make, shouldn't matter which tool makes them.

I agree that the interface between terraform and existing deployment tools seems like a generally awkward area. We've dealt with that mostly by just writing our own deployment scripts, in conjunction with what ECS provides out of the box. I'm not sure it's a problem that's solvable solely by changes to terraform, though; in this case, the fundamental problem is that there's no clean divide between the "infrastructure" part of ECS and the "application" part of ECS. That's really Amazon's fault, not terraform's.

There _is_ a clean boundary at the cluster level -- i.e. it would be easy to have terraform manage all the backing instances for an ECS cluster, and another tool manage all the services and tasks running on the cluster. If your basic philosophy is a strong divide between "infrastructure" and "applications", it seems like drawing that line right through the middle of a task definition creates much too complicated a boundary to easily manage.

**Comment by schmod** [comment]
Right.  The problem is that (in my use-case, and probably most others) an application deployment should change exactly one parameter on the task definition (`image`).  

It's difficult to draw a line around the task definition, however, because it contains a _lot_ of other configuration that I'd really prefer to remain static (and managed by Terraform).  This makes it unattractive to draw a clean boundary at the cluster level (and also leaves both your Service and Task Definition completely unmanaged by Terraform).  

As I mentioned earlier, the `ignore_changes` flag has been used elsewhere to help accommodate similar use-cases, and there's probably room to build out support for that in a way that shouldn't require any fundamental changes to how Terraform works.

**Comment by dev-head** [comment]
we share the same use case as most people are reporting here. 

our deployments are uniquely tagged, which requires a new task definition to update the ecs service on each deployment. This happens outside of the control of terraform due to a variety of reasons which are not important to the issue at hand. 

Seems like we need the ability to ignore changes on aws_ecs_service resource; we can't do that right now due to TF not supporting interpolations in lifecycle blocks as this resource is part of shared module.

**Comment by damscott** [comment]
I worked around this by using a bash script in an [External Data Source](https://www.terraform.io/docs/providers/external/data_source.html) to return the current image for the container definition. If the script gets an error looking up the task definition then it assumes this is the initial infrastructure deployment and it uses a default value.

```hcl
resource "aws_ecs_task_definition" "task" {
  family = "${var.app}-${var.env}"
  task_role_arn = "${aws_iam_role.app_role.arn}"
  container_definitions = <<JSON
[
  {
    "name": "${var.app}",
    "image": "${aws_ecr_repository.app_repo.repository_url}:${data.external.current_image.result["image_tag"]}"
  }
]
JSON
}

data "external" "current_image" {
  program = ["bash", "${path.module}/ecs-get-image.sh"]
  query = {
    app = "${var.app}"
    cluster = "${var.cluster_id}"
  }
}
```

ecs-get-image.sh:
```shell
#!/bin/bash

# This script retrieves the container image running in the current <app>-<env>
# If it can't get the image tag from AWS, assume this is the initial
# infrastructure deployment and default to "latest"

# Exit if any of the intermediate steps fail
set -e

# Get parameters from stdin
eval "$(jq -r '@sh "app=\(.app) cluster=\(.cluster)"')"

taskDefinitionID="$(aws ecs describe-services --service $app --cluster $cluster | jq -r .services[0].taskDefinition)"

# Default to "latest" if taskDefinition doesn't exist
if [[ ! -z "$taskDefinitionID" && "$taskDefinitionID" != "null" ]]; then {
  taskDefinition="$(aws ecs describe-task-definition --task-definition $taskDefinitionID)"
  containerImage="$(echo "$taskDefinition" | jq -r .taskDefinition.containerDefinitions[0].image)"
  imageTag="$(echo "$containerImage" | awk -F':' '{print $2}')"
} else {
  imageTag="latest"
}
fi

# Generate a JSON object containing the image tag
jq -n --arg imageTag "$imageTag" '{"image_tag":$imageTag}'

exit 0
```

It triggers a new task definition in Terraform when anything in the container_definition _besides_ the image is changed, so we can still manage memory, cpu, etc, from Terraform, and it plays nicely with our CI (Jenkins) which pushes new images to ECR and creates new task definitions to point to those images.

It may need some reworking to support running multiple containers in a single task.

**Comment by endofcake** [comment]
I've also used an external data source as a workaround. The main difference is that it's written in Python, supports multiple containers in the task definition, and does not fall back to `latest` (it's the responsibility of Terraform).

The script is here:
https://gist.github.com/endofcake/4ea2ac5c030a37965b65c7591c83a047

Here's a snippet of Terraform configuration that uses it:
```
data "external" "active_image_versions" {
  program = ["python", "/scripts/get_image_tags.py"]

  query = {
    cluster_name = "${data.terraform_remote_state.ecs.ecs_cluster_id}"
    service_name = "${var.app_name}"
  }
}

<...>
data "template_file" "task_definition" {
  template = "${file("task_definitions/sample.tpl")}"

  vars {
# lookup the image in the external data source output and default to 'latest' if not found
    app_image              = "${aws_ecr_repository.jenkins.repository_url}:${lookup(data.external.active_image_versions.result, var.app_name, "latest")}"
    proxy_image           = "${aws_ecr_repository.jenkins.repository_url}:${lookup(data.external.active_image_versions.result, var.proxy_name, "latest")}"
  }
}
```

This solved the problem with Terraform trying to reset the image in the task definition to the one it knew about. _However_, after an app deployment which happens outside of Terraform it still detects changes in the task definition the next time it runs. It then creates a new task revision, which triggers a bounce of the ECS service - essentially a redeployment of the same image version. I could find no way to prevent it from doing this so far.

**Comment by endofcake** [comment]
After some investigation it looks to me like the problem is caused by Terraform not knowing about the new task revision.

Say, the last revision it knows about is `36`. This is the version stored in it remote state, it's also the revision used by the ECS service as far as Terraform is concerned. The currently active revision is meanwhile `38`, and it uses a new Docker image. With the workarounds like above Terraform is able to grab this image version by describing the current ECS task definition, but it then tries to create a new task revision with it, which in turn triggers a redeployment of the ECS service.

This lack of clear separation between infrastructure and application deployments turns out rather problematic, and I'm not sure how we can work around that.

**Comment by mboudreau** [comment]
How has this not been resolved yet?  ECS has been around for a while and CI deployments outside of terraform seems like the standard operating procedure, and yet here I am still trying to get this new deployment working...

**Comment by endofcake** [comment]
See also this approach, which looks more promising:
https://github.com/terraform-providers/terraform-provider-aws/pull/3485

**Comment by codergolem** [comment]
Hi everyone,

Sorry but I am struggling to understand the problem most of people are having, namely : why do you want to avoid a new task definition revision being created when the image has changed? is not that the standard way of deploying a new image to ecs, or how are you doing it otherwise?

**Comment by endofcake** [comment]
@codergolem , it's not about avoiding the new task definition, it's about making Terraform play nicely with changes which happen outside of Terraform, namely application deployments in ECS. Terrafrom is an infrastructure management tool, and just doesn't cut it when it comes to application deployments, even if we bolt on wait-conditions on it. This really looks more like a problem with AWS API than with Terraform, but so far I see no way to resolve this impedance mismatch cleanly.

**Comment by mboudreau** [comment]
@codergolem To put @endofcake reply into context, let me provide our example:

- We have terraform build an ECS cluster, ECR, security groups, roles, services, tasks, etc.  This is to make sure our stack is solid and reproducible between dev/prod and between our different regions.
- After the cluster/services are created, we then use a CI server (in this case TravisCI) to build our docker image, push it to ECR, then use the AWS CLI to update the task definition to use the latest built image that we just pushed to ECR.

We do it this way for several reasons:
1. The code doesn't really have to know much about the infrastructure, just the ECR name, the cluster name and the task to update.
2. The terraform template doesn't have to care about the code or which version is considered the "latest" in the ECR docker image
3. It's simpler to have a single process to deploy a new version, instead of have one create the new version, then another process to update the task with the latest version.
4. Security; if you want to deploy this in a single process using Terraform, the AWS user that needs be in use on TravisCI (or any CI server) would have to have some fairly open permissions, which is a massive security vector.  I'd much rather leave the terraform apply step to be done on a developers computer where human interaction is required for having such high priviledges, then have the CI server have a user with very simple permissions (ie. ECR push, task update)

Because of these reasons, it's making it very difficult to use terraform with a CI server when you want to specify the task definition structure within terraform, which I would implore is needed since it would need references to the role created for the task or any other references being used for the infrastructure.

**Comment by endofcake** [comment]
We have a setup which is similar to what @mboudreau described (although we run Terraform on CI, not on dev machines - but the infrastructure pipeline is completely different to how we deploy the apps).

Here's a blog post about our use case https://devblog.xero.com/ci-cd-with-jenkins-pipelines-part-1-net-core-application-deployments-on-aws-ecs-987b8e032aa0.

**Comment by endofcake** [comment]
@gregharvey , this may depend on how you configured `deployment_maximum_percent` and  `deployment_minimum_healthy_percent` of your service I'm guessing?

**Comment by mboudreau** [comment]
@gregharvey  it doesn't matter, I don't want to have to taint multiple resources just for my terraform project to work every time I use it.

**Comment by menego** [comment]
This solved it for me https://forums.aws.amazon.com/thread.jspa?threadID=257234 
Basically I had some placement strategy and constraints that prevented the service to kick in the new tasks gracefully (draining the old installing the new).
In the end just adding the followings in the service definition solved the problem:

> deployment_maximum_percent = 100
> deployment_minimum_healthy_percent = 50

**Comment by Esya** [comment]
@menego This is not the issue that we're discussing here. We're discussing how we can have a nice workflow where something external from terraform can create new task definitions, without terraform destroying this task definition the next time it runs, basically.

**Comment by rjurney** [comment]
I agree that there is an unmet need here and something should be done that hasn't, although I'm not sure what that should be. Full support for the task description in terraform would be nice.

**Comment by DenisBY** [comment]
I use [ecs-deploy](https://github.com/silinternational/ecs-deploy) script to force re-deployment

**Comment by Esya** [comment]
Well on my end we've switched to terraform entreprise for a variety of reasons and that actually fixed the issue for us - as we're now changing the docker image through the TF Entreprise API, and applying directly also through the API.

**Comment by dmitrye** [comment]
Separation of Responsibilities. Terraform is trying to manage the task version of the ECS tasks that are running. Application CICD Pipeline is deploying multiple times causing the task revision to increase (something that can't be avoided with Fargate). Terraform loses site of the task version and coughs up blood. There are a couple of examples where people have bolted on scripts (python, etc) to check the current task version in ECS then set it as a variable for TF to use. Each person seems to have experienced some side effects but it's the closest thing to an answer I've seen in this thread so far.

Switching to Enterprise is great if you have tons of $$$$ but not everyone is on that path. 

While ecs-deploy is a good tool and one that I actually use in my pipeline, it's not responsible for maintaining the state of the infrastructure so that's not really an answer either.

**Comment by cwar** [comment]
With the [1.51.0 update](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1510-december-05-2018) enhancement [resource/aws_ecs_task_definition: Support resource import](https://github.com/terraform-providers/terraform-provider-aws/pull/6723) I whipped up a little script that iterates through the ECS tasks that TF knows about checks to see if there is a newer task revision via the AWS API. If there is, the script removes the task from the terraform state file and imports the newer revision.

**I want to reiterate that this script manipulates the state file via the TF CLI and is probably somewhat dangerous depending on your use case.** I'm also not very good at bash scripting. But, it works and it simplifies my workflow quite a bit. Using this in conjunction with [landscape](https://github.com/coinbase/terraform-landscape) helps me zero in on exactly what is changing when I apply the terraform plan. *Use at your own risk*.

```bash
#!/usr/bin/env bash

# get all the current TF managed ecs task defintions for the cwd state file
TASKS=$(terraform state list | grep aws_ecs_task_definition.task)

for task in $TASKS; do
    # cut out the task name
    TASK_NAME=$(echo $task | cut -d'.' -f2)
    echo "checking $TASK_NAME"

    # grep the TF revision number
    TF_REVISION=$(terraform state show $task | grep revision | grep -Eo '[0-9]{1,}')

    # get the full ARN for the ecs task without the revision
    TASK_ARN=$(terraform state show $task | grep -w task-definition | cut -d'=' -f2 | cut -d':' -f1-6)

    # get the revision of the latest active task from the aws api
    AWS_REVISION=$(aws ecs describe-task-definition --task-definition $TASK_ARN | jq .taskDefinition.revision)

    if [ $TF_REVISION != $AWS_REVISION ]; then
        echo "$TASK_NAME:"
        echo "TF_REVISION is $TF_REVISION"
        echo "AWS_REVISION is $AWS_REVISION"
        echo "does not match, removing $task from the TF state file and importing revision $AWS_REVISION"
        terraform state rm $task
        terraform import $task $TASK_ARN:$AWS_REVISION
    fi
done
```

**Comment by maartenvanderhoef** [comment]
Hi everyone, I've created a [Lambda](https://github.com/blinkist/terraform-aws-airship-ecs-service/tree/master/modules/live_task_lookup) to overcome the problem with bootstrapping. 
The lambda is configured at bootstrap and can be used as a datasource directly. It's not perfect but it works. I've explained the drift detection [here](https://airship.tf/guide/ecs_service/deployment.html#drift-detection).

**Comment by daniel-krug** [comment]
Here is a terraform-only solution similar to the scripts above.
It is lacking support for multiple images in one task-definition and also has some issue with [INACTIVE] task-definitions but maybe helps others as a starter.

Toggle bootstrapping if the service is new, afterwards set this to false.
```
variable "bootstrapService" {
  description = "set to true if you're creating a new service"
  default     = false
}
```
Fetch the current containerdefinition (image) and servicedefinition (desired_count).
```
data "aws_ecs_container_definition" "container-definition" {
  count           = "${var.bootstrapService ? 0 : 1}"
  container_name  = "${var.serviceName}"
  task_definition = "${var.serviceName}"
}

data "aws_ecs_service" "ecs-service" {
  count        = "${var.bootstrapService ? 0 : 1}"
  cluster_arn  = "arn:aws:ecs:<aws_region_here>:<your_accountid_here>:cluster/${var.clusterName}"
  service_name = "${var.serviceName}"
}
```
Make image of taskdefinition and desired_count accessible for later use.
```
locals {
  # override passed imagedefinition with currently deployed one if codepipeline has changed it in the meantime
  taskdefinitionImage = ["${var.ecrRepo}", "${data.aws_ecs_container_definition.container-definition.*.image}", "dummyentry"]

  # override passed desired_count with current desired_count if autoscaling has changed it in the meantime
  desired_count = ["${var.desired_count}", "${data.aws_ecs_service.ecs-service.*.desired_count}", "dummyentry"]
}
```
Template for the task-definition.
```
data "template_file" "task-definition" {
  template = "${file("${var.task_definition}")}"

  vars {
    image             = "${var.bootstrapService ? local.taskdefinitionImage[0] : local.taskdefinitionImage[1]}"
   <other_stuff_here>
  }
}
```
aws_ecs_service resource.
```
resource "aws_ecs_service" "ecs-service" {
  desired_count                      = "${var.bootstrapService ? local.desired_count[0] : local.desired_count[1]}"
  <other_stuff_here>
}
```

Best regards

**Comment by bploetz** [comment]
The new target sets feature (https://aws.amazon.com/about-aws/whats-new/2019/03/aws-fargate-and-amazon-ecs-support-external-deployment-controlle/  addressed  in #8131) helps solve this problem, no?

**Comment by peterdeme** [comment]
Unfortunately I join those people who need a custom script to achieve this. As the outcome became nearly perfect I'm sharing mine. It's for Windows though (PowerShell script).

I want to support 3 use cases (as 99% of the people I guess):
- App deployment where we need a new Docker image
- First time deployment (aka bootstrapping) of a task definition, the image can be a dummy one because the very first deployment will override it anyways
- Any infrastructure change that is not related to task definitions. In this case I don't want to have a new task definition.

`Lookup-TaskDefinition.ps1`
```powershell
param ($taskDefName)

$taskDef = Invoke-Expression "aws ecs describe-task-definition --task-definition $taskDefName" | ConvertFrom-Json

if ($LASTEXITCODE -ne 0 -or $null -eq $taskDef)
{
    return "{ ""image"": """" }"
}

$image = $taskDef.taskDefinition.containerDefinitions[0].image

return "{""image"": """ + $image + """}"

```

`main.tf`
```terraform
resource "aws_ecs_service" "svc" {
    name = "${var.name}-svc"
    cluster = "${var.clusterId}"
    desired_count = "${var.desired_count}"
    task_definition = "${aws_ecs_task_definition.taskdef.arn}"
}

resource "aws_ecs_task_definition" "taskdef" {
    family = "${local.familyName}",
    container_definitions = <<DEFINITION
[
  {
    "name": "${var.name}",
    "image": "${coalesce(local.imageForDeployment, data.external.lookupcurrent.result.image, "bootstrapped")}",
    "memory": ${var.mem_hardlimit},
    "memoryReservation": ${var.mem_softlimit},
    "environment": [{
      "name": "${var.environment_key}",
      "value": "${var.environment_value}"
    }],
    "essential": true,
    "portMappings": [
       {
          "protocol": "tcp",
          "containerPort": ${var.container_port}
        }
    ]
  }
]
DEFINITION
}

data "external" "lookupcurrent" {
  program = ["Powershell.exe", "${path.root}\\modules\\cluster\\scripts\\Lookup-TaskDefinition.ps1 -taskDefName ${local.familyName}"]
}

locals {
  familyName = "${var.name}-${terraform.workspace}"
  imageForDeployment = "${var.image_tag == "" ? "" : "${var.repository_url}:${var.image_tag}"}"
}

```

Important! `data.external.lookupcurrent` and `aws_ecs_task_definition.taskdef` cannot reference each other because there would be a cyclic dependency. The way I resolved it I have introduced a local variable called `local.familyName` and both the task definition and my lookup script uses it.

Running app deployment:
`terraform plan -var image_tag=[YourGitHash]`

Running non-app deployment:
`terraform plan`

Running app-deployment and infrastructural change at the same time:
`terraform plan -var image_tag=[YourGitHash]`

`coalesce` is a great function that serves me well for our use case.

**Comment by ian-axelrod** [comment]
> The new target sets feature (https://aws.amazon.com/about-aws/whats-new/2019/03/aws-fargate-and-amazon-ecs-support-external-deployment-controlle/ addressed in #8131 and #8133) helps solve this problem, no?

This feature looks wonderful, but I do not see how it solves the issue in question. If you create the service and the task set in Terraform, you will still need to create the task definition. If you have a non-Terraform process that updates that task definition, it will do so by creating a new task set, scaling the old task set down, and then updating the service's primary task set to that new task set. Terraform will see this difference, so you end up with the same issue as before.

I am going to theorize from this point onwards. Apologies in advance for the wall of text, and thank you to any that read it and comment. I want to get these thoughts out somewhere that isn't an electronic notepad or blog no one reads.

I do not believe there is a clean way to handle the issue in question without a clean separation of infrastructure and application code attributes, and that is difficult to do unless AWS alters the task definition schema. The approach that multiple folks in this thread take is to update only the image attribute, but I think that there are other task definition attributes that make sense to update outside of Terraform. What about environment variables that are meaningful only to the service, that is, variables that do not reference a Terraform-managed resource? How about docker labels? Command and entrypoint? Secrets?.... that's an interesting one.

I have trouble defining the point at which infrastructure ends for service and task definition attributes. I understand infrastructure as an ontology that does not include application code _or_ things that are meaningful only to that code. `adb.sdjfsijfzxc.us-east-2.rds.amazonaws.com` is meaningful because it refers to a database. It may only be of use to one service, but it was not designed with that in mind. However, the environment mapping `DATABASE_HOST=adb.sdjfsijfzxc.us-east-2.rds.amazonaws.com` is only meaningful to the code that interprets it. I can delete or update the mapping and the _explicit_ impact is limited to my own service. If I, as a developer, want to use the mapping `MYAPP_DATABASE_HOST=adb.sdjfsijfzxc.us-east-2.rds.amazonaws.com` instead, I should be able to do so without thinking about the impact on infrastructure. Maybe I misunderstand infrastructure itself. If so, everything I say from here onwards is invalid.

To me, a service and task as ECS-provided abstractions are infrastructure components, but the attributes of each blur the line. A service that maps to two tasks constrained to 3 EC2 instances of a certain type, with the tasks further constrained in their access to other resources (memory, databases, ...), is infrastructure. The task definition's _cpu_, _memory_, _memoryReservation_, and network setting attributes are all attributes of the task definition as an infrastructure component. These have an explicit relationship to other infrastructure components, such as the EC2 instance, though I don't think Terraform enforces these relationships. Same with task (execution) role arn. It has explicit relationships with other infrastructure components. _environment_, however, is not an infrastructure attribute, even if contains references to Terraform-managed resources. _image_ does not strike me as an infrastructure attribute. As for services, I think their task definition revision attribute is not an attribute of an infrastructure component because task definitions are 'hybrid' abstractions.

I think there should be a 'task constraints', 'task parameters', or 'task meta' concept (not task sets), and that is what services point to and Terraform manages. This entity would define resources available to individual containers and resources that all containers may share. It would place constraints on the containers, such as max memory consumption, and docker security options (e.g., no-new-privileges). What it would not do is specify what the containers _are_. During deployment, you create a task definition* (environment [which may reference resources that the task meta exposes] , image, command, anything meaningful only to your application) and connect it to the appropriate task constraints... like filling in available slots, or you just update an existing task definition without changing its attachments. This setup is more complex on the surface; however, if it correctly models my understanding of the infrastructure-service boundary better than what exists now, and it eliminates the issues in question, then it reduces complexity. At least for terraformers ;). 

Anyway, I am rambling. I am a novice Terraformer, so perhaps I have completely misunderstood the tool. Thanks again to any that actually read this.

**Comment by bploetz** [comment]
I just implemented blue/green deployments via CodeDeploy for my ECS service managed by Terraform, and just wanted to note what worked for me in case it helps someone else.

1. The `aws_ecs_task_definition` is set up to point at the image tagged "latest", such that if we ever need to destroy/re-create the resource, it's always pointed at the latest and greatest image.
2. Our CI/CD pipeline triggers a CodeDeploy deployment using the AWS CLI tools.
3. The `aws_alb_listener` resource is configured to ignore changes to `default_action`, for when the target group is switched between blue and green during a deployment.

```
resource "aws_alb_listener" "my_service" {
  lifecycle {
	ignore_changes = ["default_action"]
  }
}
```

4. The `aws_ecs_service` resource is configured to ignore changes to `load_balancer`, `task_definition`, and `service_registries` (if you're using service discovery), as these all get touched during a CodeDeploy deployment.

```
resource "aws_ecs_service" "my_service" {
 deployment_controller {
    type = "CODE_DEPLOY"
  }

  lifecycle {
    create_before_destroy = true
	ignore_changes = ["load_balancer","task_definition","service_registries"]
  }
}
```

Once I made these changes on the Terraform side, if I run `terraform plan` after a CodeDeploy deployment, Terraform says there's nothing to do for the service in question.

HTH.

**Comment by beephotography** [comment]
@bploetz That really sounds like a good solution, but I see one pitfall though: the `terraform.tfstate` file is not updated to the new revisions. You may run into trouble if (in some situation) terraform **needs** to know the current revision (etc.). What do the others think, are my concerns reasoned? I'm really not sure :thinking:

**Comment by bploetz** [comment]
@beephotography Yup, that's basically the decision you're making with this approach: Terraform can create/destroy resources, but after creation they belong to someone else to manage. I can't think of any situations where the out of date state file would be a problem given the resources in question, but even if I needed to change something fundamental about the set up such that I needed an up to date state file, at that point I would likely just bring up new resources in parallel with the changes, and then deprecate/destroy the old resources.

I may be missing something obvious with this approach, and if so, I'm sure someone will set me straight. :)

**Comment by jfirebaugh** [comment]
@endofcake Thanks for outlining the central issue:

> Say, the last revision it knows about is 36. This is the version stored in it remote state, it's also the revision used by the ECS service as far as Terraform is concerned. The currently active revision is meanwhile 38, and it uses a new Docker image. With the workarounds like above Terraform is able to grab this image version by describing the current ECS task definition, but it then tries to create a new task revision with it, which in turn triggers a redeployment of the ECS service.

Everything else can be worked around with strategic use of `data "aws_ecs_task_definition"` and `data "aws_ecs_container_definition"`, but I cannot figure out how to get terraform not to force creation of useless new task revisions.

Your [blog post](https://devblog.xero.com/ci-cd-with-jenkins-pipelines-part-2-managing-infrastructure-with-terraform-and-docker-bd4b81554aa7) doesn't seem to explain how this is solved either. Did you figure it out?

**Comment by alevinetx** [comment]
Last post was in April, and it's now October.

   Has anyone found a viable solution to this, short of using external scripts to update-template-in-place ?      I've tried using the lifecycle ignore_changes "*" on ecs_task_definition, but either that's not far reaching enough ( ```"As an example, this can be used to ignore dynamic changes to the resource from external resources. Other meta-parameters cannot be ignored."``` ), or it's not fully implemented;  I wouldn't quite know where to look to see.

   FWIW, we're still on .11.   If moving to .12 will help us reach this goal, then that's what we'll prioritize.

Any insight is greatly appreciated.

**Comment by jfirebaugh** [comment]
@alevinetx We recently migrated from 0.11 to 0.12. Lots of nice improvements, but unfortunately this issue isn't one of them.

**Comment by riwiki** [comment]
After researching a lot, I found the following to be a promising way within terraform alone. The goal: if terraform scripts are executed locally, leave the image version/tag unchanged (thus not updating the ECS service's task definition), but allow to specify a specific (new) image version to be deployed which then updates the task definition, e.g. to be used in CI/CD or when rolling back to a specific version manually.
```
variable "app_version" {
  type = number
  default = 0
}

data "aws_ecs_task_definition" "active" {
  task_definition = local.task_family
}

data "aws_ecs_container_definition" "active" {
  task_definition = data.aws_ecs_task_definition.active.id
  container_name = "containername"
}

locals {
  task_family = "task name"
  # digest is the image's version/tag
  new_version = var.app_version > 0 ? var.app_version : data.aws_ecs_container_definition.active.image_digest
}

resource "aws_ecs_service" "apps-service" {
  task_definition = "${aws_ecs_task_definition.task.arn}"
  ...
}

resource "aws_ecs_task_definition" "task" {
  family                = local.task_family
  container_definitions = "${data.template_file.container_definitions.rendered}"
}

data "template_file" "container_definitions" {
  template = file("${path.module}/templates/task-definition.json")
  vars = {
    app_version = "${local.new_version}"
  }
}
```

This does not cover that initially no task definition is present -- as I have no stage to test it currently -- but maybe someone could add sufficient logic. Also, I am not sure what terraform version this requires. If you have version >= `0.12.7` you can use the `regex` function to handle alphanumeric image digests. I hope it gives you all necessary pointers. Otherwise, please poke me and I will elaborate a bit.

### Approaches that did not work, and why

* add lifecycle `ignore_changes` conditionally. No conditions are allowed within lifecycle.
* add `overrides.tf` to mask the `ignore_changes` section. Cannot access resources that are nested within a module. Also, needs some cumbersome file management to move/unmove the `overrides.tf` file depending on use case.
* external data source; although feasible, requires dependencies, e.g. for parsing json. Seems rube-goldberg-machinish.
* work with `terraform taint`. Add a `ignore_changes` block to the task definition resource regarding its `container_definitions` and have the resource built with new versions by tainting it beforehand. This introduces a race condition!! So deploying a new version or not deploying a new version solely depends on whether a taint operation was performed beforehand -- and you can only hope that the very same context had executed it deliberately. So better not use it. Also, I need to define the resources to be tainted anew.

## Side note
I was somehow confused in this conversation. Some people seem to only want to make sure that terraform updates the ECS service definition with the latest task definition revision when that task definition is changed. For me, terraform 0.12 handled this just fine: whenever I changed the task definition, the resource needed to be replaced and the service needed to be updated because of the task-resource's arn would change. I performed deploys only via terraform, so there is e.g. no manual deployment via the AWS console. Maybe that was the aspect I missed.

**Comment by Sytten** [comment]
The problematic code for us is really this line: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_ecs_task_definition.go#L417

**Comment by nickdani** [comment]
**I'm facing the same issue.** 

**In short:**
- I have use terraform only for creating/deleting resources
- I use TravisCI + ecs-deploy to build and deploy new image
- In ECS I keep only a few task definitions

**The problem:**
after a few deployments the old task definitions created by terraform became INACTIVE and terraform tries to create a new task definition.

Based on all the comments I don't see any elegant solution for that, except having a flag "bootstrap" - which became inactive as soon as you created a task definition at the beginning, although you still have a problem if you need to update task definition in terraform(for example add env variables) - and then it become messy.

I agree with @Sytten and I think it would be great if it would be possible to mark somehow task definition to use the last revision or mark resource like "create_once".

As some people pointed out, current implementation simply makes impossible to use Terraform + ECS + CI/CD

**Comment by sworisbreathing** [comment]
@Sytten - I don't know if that's really the problem. Stripping out the inactive task definitions from the read result kind of makes sense, because per https://docs.aws.amazon.com/AmazonECS/latest/userguide/deregister-task-definition.html
> You cannot use an `INACTIVE` task definition to run new tasks or create new services, and you cannot update an existing service to reference an `INACTIVE` task definition

I'd definitely prefer that Terraform not try to update an ECS service to point to an inactive task definition, so I'm kind of okay with that behavior.

What I see as more of a problem, is two-fold:
1. When Terrraform decides that a task definition needs changing (really a new revision in the task family), it insists on de-registering the old task definition. Whether that happens first or second (i.e. `create_before_destroy`) is moot - deregistering the existing task definition is still mandatory.
1. When something external to Terraform (e.g. my CI/CD pipeline) has rolled out a new version and updated the ECR image, ECS task definition, and ECS service to all be what they are supposed to be, Terraform sees that as a reason to change the ECS task definition (thus triggering point number one), even when things are already how I want them. It's simply a consequence of Terraform's state file being out-of-sync with the infrastructure, and the fact that ECS doesn't have an "update" API that will transparently create a new revision of an existing task definition.

What I want Terraform to do is:
1. If the task family doesn't exist, create the task definition and service
1. If the service is already pointing to the latest task definition, which is already pointing to the latest image in ECR, do nothing.
1. If a newer image has been pushed to ECR, but the above wasn't met, then terraform should create a new revision of the task definition and update the service to point to it, but the existing taskdefinition should be left alone (i.e. don't deregister it).

So, all that being said, I think a solution would be something along the lines of:
* Remove `ForceNew: true` from task definition attributes other than `family`
* In resourceAwsEcsTaskDefinitionUpdate, add a call `RegisterTaskDefinition` when a change needs to be made that isn't just tags
* Introduce a boolean flag on the `ecs_task_definition` attribute, to switch between the existing behavior and the new one (probably safest to default to the old behavior)

**Comment by Sytten** [comment]
@sworisbreathing I don't see how my comment is contradictory to yours. Terraform should always compare against the latest version instead of abandoning and just saying the resource needs to be created if the current task definition is inactive.

**Comment by sworisbreathing** [comment]
@Sytten fair point - if your terraform templates declare an ecs task definition resource, and an ecs service resource that depends on the task definition, and the task definition is deregistered, nothing should change as far as Terraform is concerned, because nothing has changed as far as ECS is concerned - the service will continue to run against the task definition, and can even scale up or down using the inactive task definition. So, yeah, maybe in that scenario Terraform should only view an inactive task definition as requiring a change, if the service which depends on it is marked as needing to be changed.

Is that even possible for Terraform internals to figure out?

**Comment by Sytten** [comment]
Exactly! It is possible to figure that out, thats how other tools (like ecs-deploy) do it. There is a bit of AWS API magic and comparison, but its doable.

**Comment by sworisbreathing** [comment]
What I mean is, does terraform (or terraform-plugin-sdk) provide the ability to mark resources as requiring a change, in reverse dependency order?

Ordinarily, terraform would work out that the ECS service needs to be changed because it depends on an ECS task definition which is changing. But in your case, it's a bit of a chicken-and-egg scenario. The inactive task definition only needs to be marked as changed (actually a new one created) because the ECS service (which depends on it) has some other attribute that needs changing.

**Comment by sworisbreathing** [comment]
The PR I opened delivers this feature for my use case. A bit of info about my setup:

Our CI/CD pipeline applies 2 tags to the image it builds:
* a permanent tag following the pattern `<semver>-<build_timestamp>-<short_git_hash>` (e.g. `1.0.0-20200110115259-ac346af`)
* a transitive tag - `active`

Note that because the permanent tag always starts with a numeric character, we will get consistent ordering when we call `sort(data.aws_ecr_image.app.image_tags)`.

When CI/CD runs, it will:
* build and tag the image according to our tagging strategy
* push the image/tags to ECR
* register a new revision of the task definition, pointing at the new permanent tag
* point the ECS service to the new revision of the task definition

The Terraform templates set up the ECS task definition and service. For the task definition, we will use the permanent tag of whichever ECR image is tagged as "active". Our terraform (0.11) templates look like this:

```hcl
data "aws_ecr_repository" "app" {
    name = "app"
}

data "aws_ecr_image" "app" {
  registry_id = "${data.aws_ecr_repository.app.registry_id}"
  repository_name = "${data.aws_ecr_repository.app.name}"
  image_tag = "active"
}

resource "aws_ecs_task_definition" "app" {
  family = "app"
  container_definitions = <<EOF
    [
        {
            "image": "${data.aws_ecr_repository.app.repository_url}:${element(sort(data.aws_ecr_image.app.image_tags), 0)}"
            ...
        }
    ]
EOF
  ...
}

resource "aws_ecs_service" "app" {
    task_definition = "${aws_ecs_task_definition.app.family}:${aws_ecs_task_definition.app.revision}"
    ...
}
```

Now, to show a basic workflow, consider the following scenario.

We'll start off with Terraform being in sync with what's in ECS:

```bash
# Show what the actual latest active task def revision is
$ aws ecs describe-task-definition --task-definition app --query 'taskDefinition.revision'
1

# Show what Terraform thinks the current task def revision is
$ terraform state show aws_ecs_task_definition.app | grep 'revision'
revision                                        = 1



# Show the actual task definition the service is using
$ aws ecs describe-services --cluster app --services app --query 'services[0].taskDefinition'
"arn:aws:ecs:<region>:<account>:task-definition/app:1"

# Show which task def revision Terraform thinks the service is using
$ terraform state show aws_ecs_service.app | grep task_definition
task_definition                           = app:1
```

After CI/CD has run, our Terraform state is now out of sync with ECS:

```bash
# Show what the actual latest active task def revision is
$ aws ecs describe-task-definition --task-definition app --query 'taskDefinition.revision'
2

# Show what Terraform thinks the current task def revision is
$ terraform state show aws_ecs_task_definition.app | grep 'revision'
revision                                        = 1



# Show the actual task definition the service is using
$ aws ecs describe-services --cluster app --services app --query 'services[0].taskDefinition'
"arn:aws:ecs:<region>:<account>:task-definition/app:2"

# Show which task def revision Terraform thinks the service is using
$ terraform state show aws_ecs_service.app | grep task_definition
task_definition                           = app:1
```

With the changes I made in my PR, here's what happens when I run `terraform apply` 

```bash
$ terraform apply
...
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
...

# Show what Terraform thinks the current task def revision is
$ terraform state show aws_ecs_task_definition.app | grep 'revision'
revision                                        = 2

# Show which task def revision Terraform thinks the service is using
$ terraform state show aws_ecs_service.app | grep task_definition
task_definition                           = app:2
```

**Comment by gpkc** [comment]
I agree that the container definition should be a 1st-class citizen similar to policies, this way we could simply have a lifecycle to ignore in case it changes. It is very weird that we can't update some other configuration on the container definition, such as an environment variable, without having to replace the image with an old one - or having to change it as well. As others said, Terraform shouldn't mess with application deployment, and the way to update ECS applications is by changing task definitions. There should be a way to accommodate that.

**Comment by medcross** [comment]
I came up with another solution to overcome that problem. It builds on the idea of using a _aws_ecs_task_definition_ data block (https://www.terraform.io/docs/providers/aws/d/ecs_task_definition.html)

### Prerequisites
You have an existing ECR repository containing a Docker image tagged latest. The image could be pushed by yourself or an existing deployment pipeline.

### Terraform Configuration

1. Create a new Terraform working directory for ECS resources only. Resources like Secrets from SecretsManager can be accessed via the _remote_state_ data block.

2. Get the latest image by using the _aws_ecr_image_ data block
```hcl
data "aws_ecr_image" "latest" {
  repository_name = var.ecr_repository_name
  image_tag       = "latest"
}
```

3. Write every configuration value of your desired task definition in a _local_ block. If wrting a reusable module, some values could also be provided via variables.

4. Write a _container_definitions.json.tmpl_ template with a variable image name. Update the _local_ block to include the variable map of the _templatefile_ function.
```hcl
locals {
  latest_image = "${data.aws_ecr_image.latest.repository_name}:${data.aws_ecr_image.latest.image_tag}"

  container_definition_vars = {
    name           = var.service_name
    container_port = var.container_port
  }
  container_definition_bootstrap_vars = merge(local.container_definition_vars, {
    image = local.latest_image
  })

  task_definition_requires_capabilities = ["FARGATE"]
}
```

5. Write a "bootstrap" _aws_ecs_task_definition_ resource. This resource may not be updated by Terraform after its creation.
```hcl
resource "aws_ecs_task_definition" "bootstrap" {
  family                   = var.service_name
  container_definitions    = templatefile(var.container_definitions_path, local.container_definition_bootstrap_vars)
  cpu                      = var.container_cpu_limit
  memory                   = var.container_memory_limit
  requires_compatibilities = local.task_definition_requires_capabilities
  tags                     = var.default_tags

  lifecycle {
    ignore_changes = all
  }
}
```

5. Include the following two data blocks to retrieve information about the active task definition. You can pass the active task definition to a service.
```hcl
data "aws_ecs_task_definition" "active" {
  task_definition = aws_ecs_task_definition.bootstrap.family
}

data "aws_ecs_container_definition" "active" {
  task_definition = aws_ecs_task_definition.bootstrap.family
  container_name  = var.service_name
}
```

6. To allow applying new configurations for i.e. CPU limits a second task definition is required. Copy and paste the bootstrap _aws_ecs_task_definition_ resource block. Ignore changes to i.e. the family_name to prevent errors later on.

7. Change the image name of the templatefile by using another _local_. Use the image of the active task definition's container definition.
```hcl
locals {
# Other locals ...
  container_definition_actual_vars = merge(local.container_definition_vars, {
    image = data.aws_ecs_container_definition.active.image
  })
}
```

**Comment by siassaj** [comment]
I'd love a fix for this however my solution, which works except for the first time setting up an environment, is:

```
data "aws_region" "current" {}
data "aws_caller_identity" "current" {}
data "aws_ecs_task_definition" "default" {
  task_definition = local.task_definition_family
}
data "aws_ecs_container_definition" "www" {
  task_definition = local.data_task_definition_arn
  container_name  = local.task_definition_container
}

locals {
  task_definition_family = var.environ_name
  task_definition_container = "www"
  data_task_definition_arn = "arn:aws:ecs:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:task-definition/${local.task_definition_family}:${data.aws_ecs_task_definition.default.revision}"
  docker_image = data.aws_ecs_container_definition.www.image
}
```

From there you can use `local.docker_image` in your container template. I use
https://github.com/cloudposse/terraform-aws-ecs-container-definition.

There is still the problem however that terraform may try to create a new task definition with an old revision number rather than incrementing the most recent revision number.

It's cosmetic, since it'll diff the task definition & set all the values to the most recent, but it's still no bueno i guess.

Anyway, this works both on local machine & on terraform cloud, where I couldn't get an external provider to work because I couldn't get AWS CLI to work on their runner environments, despite the docs saying that its' latest Ubutnu LTS.

**Comment by siassaj** [comment]
> which works except for the first time setting up an environment

Adding more fuel on this dumpster fire, we need to be able to spin up & down dev swim lanes at leisure so have had to do:

```
variable "is_first_apply" {
  type = bool
  default = false
}

data "aws_region" "current" {}
data "aws_caller_identity" "current" {}

data "aws_ecs_task_definition" "graphql_api" {
  count = var.is_first_apply ? 0 : 1

  task_definition = local.graphql_api_task_definition_family
}

data "aws_ecs_container_definition" "hasura" {
  count = var.is_first_apply ? 0 : 1

  task_definition = local.derived_graphql_api_task_definition_arn
  container_name  = local.hasura_container_name
}

locals {
  derived_graphql_api_task_definition_arn = "arn:aws:ecs:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:task-definition/${local.graphql_api_task_definition_family}:${local.derived_revision}"

  derived_revision = var.is_first_apply ? 1 : data.aws_ecs_task_definition.graphql_api[0].revision

  hasura_image  = var.is_first_apply ? "${aws_ecr_repository.hasura.repository_url}:latest" : data.aws_ecs_container_definition.hasura[0].image
}
```

So at least we can build up & tear down everything but ... this stuff is no bueno

**Comment by sworisbreathing** [comment]
@siassaj are you saying what you've done above works without manually creating an ECS task definition before running Terraform?

If so, that's awesome!

**Comment by maartenvanderhoef** [comment]
@siassaj @sworisbreathing you can also use a self-made Lambda as datasource as it's done here: https://github.com/blinkist/terraform-aws-airship-ecs-service/blob/dbc1124e86198447a9732c1c4fce84e98abacad0/modules/live_task_lookup/main.tf#L38 This way you can both bootstrap from scratch and follow up with new deployments without any needed change.

**Comment by siassaj** [comment]
@sworisbreathing yeah you just need to pass `is_first_apply = true`
@maartenvanderhoef i might have to bite the bullet & switch to that, or end up seeing if pulumi is willing to be less... restrictive. Next hurdle is getting dev & prod to have _slightly_ different architecture ...
 
It somehow feels very very wrong to have to use a lambda to avoid a 'data' resource from throwing an error :(

**Comment by maartenvanderhoef** [comment]
@siassaj of course Pulumi is less restrictive, Boto is also less restrictive, but not sure if that's the right direction tbh.
It seems https://github.com/terraform-providers/terraform-provider-aws/pull/8652 is merged, this way you can set the EXTERNAL deployment controller implying Terraform has no control anymore over the task definition.

**Comment by robertomoutinho** [comment]
Now that the "deployment_controller = EXTERNAL" was added to the aws provider (aws_ecs_service), did anyone got around the problem ?

**Comment by nugroho-expereo** [comment]
We just upgraded to terraform 0.13 and we can now use "depends on" nicely for aws_ecs_task_definition data source and it works for both first creation or update. 

```
data "aws_ecs_task_definition" "latest_task_definition" {
  task_definition = aws_ecs_task_definition.service.family

  depends_on = [aws_ecs_task_definition.service]
}

locals {
  latest_task_definition = "${aws_ecs_task_definition.service.family}:${max(aws_ecs_task_definition.service.revision, data.aws_ecs_task_definition.latest_task_definition.revision)}"
}


resource "aws_ecs_service" "service" {
  name            = var.name
  cluster         = var.ecs_cluster_id
  task_definition = local.latest_task_definition
...
}
resource "aws_ecs_task_definition" "service" {
  family                = var.name
...
}
```
With terraform 0.12 we had this issue: https://github.com/hashicorp/terraform/issues/11806  and as workaround we used an external bash script to get latest task definition, no it is not needed anymore with terraform 0.13.

To make it works with CI/CD (we use ecs deploy) on deployment we add `--no-deregister` option if latest task definition was created by terraform (we use tag to differentiate if task definition is created by terraform or ecs deploy) to avoid task definition removal that will trigger terraform change.

**Comment by siassaj** [comment]
@nugroho-expereo I think the issue that we have is that we do not use `latest` tag at all. We strictly version with semver & each task definition uses strict image version tags.

So in this case, we don't really care what task definition revision is being consumed by the service, and are fine with terraform or circleci building them as long as the image tags are correct. I'll try your `depends_on` approach & see if that works for our usage.

**Comment by briancurt** [comment]
@siassaj Have you had the chance to try it? I have a similar use case as yours and am curious about your results before trying it myself.

**Comment by ianwestcott** [comment]
I have another workaround approach to throw into the pile: 

In our setup, similar to others here, we want to explicitly declare the application version to deploy, meaning we need to be able to pass a unique tag as part of the container image on every deploy. We also want to manage infrastructure resources with Terraform, but handle deployments separately via CI. 

The solution we came up with was to have Terraform create a "template" task definition which declares all of the desired configuration for the container, but with `latest` as a placeholder container image version. The ECS service configuration uses this template task def upon creation, but has a `lifecycle` hook set to `ignore_changes` on the `task_definition` attribute. On deployment, our CI tooling reads the configuration from this "template" task definition, and uses it to create a _separate_ task definition for use by the service, injecting the desired image tag to deploy. The new task definition is not managed by Terraform. With this approach, any changes made by Terraform will apply only to the "template" task definition, and won't try to manipulate the running tasks. Such changes will be picked up by the service on the next deployment triggered by CI. 

I had hoped to use [ecs-deploy (Python)](https://github.com/fabfuel/ecs-deploy) or [ecs-deploy (Shell)](https://github.com/silinternational/ecs-deploy) to handle the deployment part, but neither seems to have working support for reading configuration from a separate task definition (see [here](https://github.com/fabfuel/ecs-deploy/issues/144) and [here](https://github.com/silinternational/ecs-deploy/pull/215), respectively) so we rolled our own method using the `aws` CLI and `jq` for now.

**Comment by siassaj** [comment]
@briancurt 

> @siassaj Have you had the chance to try it? I have a similar use case as yours and am curious about your results before trying it myself.

I didn't get far because we use specific docker image tags rather than latest. So what happens is that we need to extract that information out of the latest running task definition, then inject that into the new task definition that terraform is building, along with any changes to other things like env vars etc.

The problem comes because building a new task definition due to any changes other than the image tag causes a new build, not a reuse of the current task definition.

So we're still doing https://github.com/terraform-providers/terraform-provider-aws/issues/632#issuecomment-661685848
It makes me uncomfortable.

@ianwestcott So your approach requires CI to redeploy if you change things like env vars right? Honestly it might not be the worst way to do it if you can get terraform to cause your CI tool to redeploy. Unfortunately for us we're running on Terraform Cloud which is not so nice for running your own scripts/binaries... Maybe I need to write a provider that just launches CI redeploys when something important changes...

**Comment by ianwestcott** [comment]
> Your approach requires CI to redeploy if you change things like env vars right?

Yep! But one of our goals was to keep the process of code deployment completely separate from infrastructure management, so not having Terraform trigger a deployment is something of a feature in our case.

**Comment by sebastian-stephan** [comment]
I like your approach @ianwestcott . In case someone uses GitHub Actions, I wrote a very simple [action](https://github.com/modum-io/aws-ecs-deploy-service-by-template) that does exactly the steps you described.

**Comment by blytheaw** [comment]
I am also facing this problem. I'm intrigued by your approach, @ianwestcott, and will probably try it. We manage a large number of very similar services, so often infrastructure changes made in Terraform need to be applied to all of the services.

I'm curious how many services you are managing with this approach, and how you handle triggering redeployments for a many services after making changes in Terraform. Currently we can do this because Terraform is triggering deployments on infrastructure changes, but this suffers from all the problems mentioned in this thread.

With this suggested approach and our current tooling, we would have to manually redeploy every service after making an infrastructure change in Terraform that affects all of them. Has anyone found a good workaround that both decouples the infrastructure changes and deployment/CI but also makes it easy to make mass changes to many services easily?

**Comment by adamlc** [comment]
@blytheaw I'm wondering if you could use a [null_resource with a provisioner](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource) to trigger the changes somehow? 

I have a similar problem, just trying to find the best solution. We use GitLab so I'm going to try if I can get the null_resource to trigger a GitLab pipeline when something changes with by using a provisioner.

**Comment by bilbof** [comment]
We also hit this problem.

I agree with what others have said: ECS could make it easier to draw a line between infrastructure and application deployments. It has led to us doing something quite idiosyncratic where we'd prefer to follow convention.

I think HashiCorp's answer to this may be the recently announced [Waypoint](https://www.waypointproject.io/). Right now Waypoint doesn't integrate enough with Terraform to meet our needs (i.e. task definitions reference secrets and resources created by Terraform).

There are a few ways we explored to fix this:

a) Deploy with Terraform, use `var=app_image_tag=release_123` when applying to release specific image
b) Manage task definitions outside of Terraform, use the AWS CLI to update an ECS service created by Terraform (with `ignore_changes` set on the Service resource task_definition attribute) to use the newly minted task definition during deploys
c) Create task definitions and services in Terraform during bootstrapping, but manage task definition revisions in a separate statefile (using the same task def modules) to enable more involved / custom deploys.

We've gone for option C for now.

This gives us a single source of truth for config (apart from docker `image_tags`, which are passed in as `-var` args). This way we can have all infrastructure, including task definitions, defined in Terraform (having cake), and also update the ECS Service to use the new task definitions using the AWS CLI (eating cake too).

From an operator perspective the workflow looks like this:

```
# updating infra
cd environment && terraform apply -var=app_a_image_tag=tag_of_thing_already_in_env ...

# deploying app
cd apps/app_a
terraform apply -var=app_a_image_tag=release_123
task_definition_arn=$(terraform output task_definition_arn)
aws ecs update-service --cluster my-cluster --service app_a \
--task-definition $task_definition_arn --region $AWS_REGION
```

To add another benefit to deploying apps separately to infrastructure updates: deploy hooks. Sometimes you want to run a task, using the same task definition you use for apps, but for a different service (e.g. db migrations pseudocommand: `aws ecs run-task --task-def=my-task --cmd=rails db:migrate --service=migrator`). Things like Capistrano give you flexibility to do that kind of thing, and this approach gives you something like that (e.g. 1) create task def, 2) run tasks using def, 3) update service).

**Comment by sp3c1** [comment]
@bilbof so, when you have 40 services in this cluster, you do this manually for every service? That is a bit of nonsense.

**Comment by bilbof** [comment]
@sp3c1 thanks for your feedback. The procedure I described above is automated using a Concourse pipeline; releases are continuously deployed without manual intervention. Since you mention it… our design has changed to following procedure, which gives us more control over release deployments (this is automated):

a) ECS Services and other resources are managed by Terraform
b) Task definition revisions are pushed using the AWS CLI

I’d be keen to improve on this. If there is a generally accepted pattern for continuously deploying to ~100 ECS Services managed with Terraform, I'd like to adopt it.

**Comment by schmod** [comment]
> I’d be keen to improve on this. If there is a generally accepted pattern for continuously deploying to ~100 ECS Services managed with Terraform, I'd like to adopt it.

After 4 years of following the consensus seems to be that, no there is not.  You're (more or less) following the same compromise that most folks have landed on.

**Comment by maartenvanderhoef** [comment]
For a little while I managed one of the larger Terraform modules for ECS. Since EKS came, many issues stayed unresolved and it was clear that AWS is letting ECS for what it is. A big shame because ECS is the 'exoskeleton' way of managing docker services and from the beginning it was very close to perfection. As with many AWS products; ElasticBeanstalk, Amplify, they get to a level where it's working good enough but AWS never gives them the final paint job.

I haven't touched ECS the recent years, but my belief now is to really integrate the creation and updating of services within the CI/CD itself. SSM could be used with Terraform to centrally orchestrate parts like memory consumption which can later be used by CI/CD et cetera. This, or completely use Terraform to build up Codepipeline/Codebuild and have control over the ECS Services' configuration by managing its CICD layer.

**Comment by zen0wu** [comment]
After reading through the thread and much thinking, I decided to take the following approach.

- Have an application script to generate the task definition as the single source of truth, because our application uses TS, it's more naturally integrated with JSON and it's strongly typed (while TF template is not great for either)
- Have a `null_resource` executing the script to "ensure the task definition exists" in ECS
- Have a data source that selects the most recent active task def, since typically deployment uses the most recent active revision (at least in our case), and this depends on the null_resource

```tf
resource "null_resource" "task_definition_generator" {
  triggers = {
    family  = var.family
    command = var.generator_command
  }

  provisioner "local-exec" {
    command     = var.generator_command
    working_dir = local.root_dir
  }
}

data "aws_ecs_task_definition" "task_def" {
  depends_on = [null_resource.task_definition_generator]

  # This pulls in the latest ACTIVE revision. It might or might not be
  # the one created by the generator_command, but that's generally ok.
  # We're just assuming the latest version is always working.
  task_definition = var.family
}
```

**Comment by WhyNotHugo** [comment]
@zen0wu How do you apply updates to task definitions? E.g.: chaging the `memoryReservation`, `environmentFiles` or other attributes?

**Comment by zen0wu** [comment]
@WhyNotHugo all the actual task definition belongs to our TS code, so if we want to update those, we'll just trigger a deploy, by first calling the same generator command (to create the new task definition) and then call UpdateService.

TF only manages limited info, which includes, which load balancer to use, how many containers in total, essential things that only belongs to the service.

**Comment by WhyNotHugo** [comment]
Task definitions have a few terraform-managed resources in my case (group log, environment variables, ssm arns, and IAM arns).

How do you get those values from terraform into the deploy process?

**Comment by zen0wu** [comment]
@WhyNotHugo good question - ideally we can pass those as arguments into the generator command, by passing those in as command line arguments when terraform calls it, since TF has those values. But for the continuous delivery part (TypeScript in our case, running alone for the deploy), they'll either have to pull the values out of a - existing values in the task definition, b - from terraform (by running terraform console), or c - hard code (this is what we did :p, since we only need taskRoleArn)

**Comment by bholzer** [comment]
> Task definitions have a few terraform-managed resources in my case (group log, environment variables, ssm arns, and IAM arns).
> 
> How do you get those values from terraform into the deploy process?

First make an API call to describe the existing task definition which contains those fields. Then modify just the image (or any other fields you want) from that response, and pass it into a task def registration call. How I handled this:

```bash
task_def_family_name=${1:-}
image=${2:-}
container_name=${3:-}

if [[ -z "$task_def_family_name" ]] ; then
    echo 1>&2 "error: A task definition family name is required. May also include a revision (familyName:revision)"
    exit 1
fi

if [[ -z "$image" ]] ; then
    echo 1>&2 "error: An image is required"
    exit 1
fi

# Format the response in a way that is easily usable by register-task-definition
latest_task_definition=$( \
    aws ecs describe-task-definition \
        --include TAGS \
        --task-definition "$task_def_family_name" \
        --query '{  containerDefinitions: taskDefinition.containerDefinitions,
                    family: taskDefinition.family,
                    taskRoleArn: taskDefinition.taskRoleArn,
                    executionRoleArn: taskDefinition.executionRoleArn,
                    networkMode: taskDefinition.networkMode,
                    volumes: taskDefinition.volumes,
                    placementConstraints: taskDefinition.placementConstraints,
                    requiresCompatibilities: taskDefinition.requiresCompatibilities,
                    cpu: taskDefinition.cpu,
                    memory: taskDefinition.memory,
                    tags: tags}' \
)

container_count=$(jq -r '.containerDefinitions | length' <<< "$latest_task_definition")

if [[ "$container_count" -gt 1 ]] && [[ -z "$container_name" ]] ; then
    echo 1>&2 "error: The task definition has more than one container definition, you must choose one."
    exit 1
fi

# If there's only one container in the task definition, update its image, otherwise look by container name
# We should never make it to the `else` here, but we just create a duplicate revision if we do.
new_task_definition=$(echo "$latest_task_definition" \
    | jq -rc --arg containerName "$container_name" --arg newImage "$image" \
        '.containerDefinitions |= (
            if ( . | length ) == 1 then
                .[].image = $newImage
            elif ($containerName | length) > 0 then
                map(select(.name == $containerName).image = $newImage)
            else
                .
            end
        )'\
)

registration_response=$(aws ecs register-task-definition --cli-input-json "$new_task_definition")

new_revision=$(jq '.taskDefinition.revision' <<< "$registration_response")
old_revision=$((new_revision-1))

deregistration_response=$(aws ecs deregister-task-definition --task-definition "$task_def_family_name:$old_revision")
```

**Comment by anGie44** [comment]
Hi @dennari and all those following this issue 👋 .Thank you again for submitting/providing feedback to this issue. As noted by others in the comments above, because Terraform expects full management of the ECS Task Definition resource and the upstream ECS API does not support methods to appropriately manage individual revisions without replacement, we cannot provide any patches to the current state of resource’s behavior in the provider and thus will be closing this issue.

Patches that suggest setting or manipulating the state difference of an ECS Task Definition’s’ revision would imply having one resource responsible for multiple AWS resources, and this would prove problematic to traditional practitioner experience with the rest of the Terraform provider ecosystem. With that said, we want to note that the ECS API does not allow for a great user experience when using Terraform and separate tooling like `ecs-deploy`, and thus encourage practitioners to proceed with caution when integrating both in a CI/CD pipeline.

**Comment by WhyNotHugo** [comment]
@bholzer My issue with that approach is that the next time terraform runs, it detects changes in the task definitions, and tries to recreate them.

I used to have a hack to work around this: a `data aws_ecs_container_definition` which found the latest version and re-used the image defined there. However, this had two issues:

- Each time there's a deployment, terraform will see the dirty state and update the task definitions. The newly created one matches the latest anyway, but it adds a lot of diff noise each time terraform is run, especially with a lot of task definitions.
- Creating a new task definition is a pain due to the circular dependency between the `data` and `resource`.

I finally found a solution that really works from all angles. I create a "template" task definition in terraform, which is fully terraform-managed and never altered outside of terraform:

```hcl
resource "aws_ecs_task_definition" "django_template" {
  for_each = local.full_websites

  family = "django-${each.key}-template"
  container_definitions = jsonencode([{
    name              = "django"
    command           = ["/app/scripts/run-django"]
    essential         = true
    image             = "whynothugo/sleep"
    memoryReservation = 600
    portMappings      = [{ containerPort = 8000, protocol = "tcp" }]
    user              = "django"

    # Zeep's cache failes with this on:
    # readonlyRootFilesystem = true
    linuxParameters = { tmpfs = [{ containerPath = "/tmp", size = 100 }] }

    environmentFiles = [
      {
        value = local.envfile_arns[each.key]
        type  = "s3"
      }
    ]

    healthCheck = {
      command  = ["/app/scripts/healthcheck-django"]
      interval = 30
      retries  = 3
      timeout  = 5
    }

    logConfiguration = {
      logDriver = "awslogs"
      options = {
        awslogs-group         = aws_cloudwatch_log_group.django[each.key].name
        awslogs-region        = "us-west-2"
        awslogs-stream-prefix = "ecs"
      }
    }

  }])

  task_role_arn      = aws_iam_role.production_task_role.arn
  execution_role_arn = aws_iam_role.production_task_execution_role.arn
  network_mode       = "bridge"

  placement_constraints {
    type       = "memberOf"
    expression = "attribute:Role == Web"
  }

  requires_compatibilities = ["EC2"]

  tags = {
    Component    = "Django",
    Environment  = "Production"
    BaseImageUrl = aws_ecr_repository.production.repository_url
  }
}
```

These don't have the right image though. My deployment pipeline will find the task definition (the one with `-template`), replace the image, and save it under a new name (without `-template`). My ECS services point to the non-template ones.

The base image URL for each service is specified in the tags, so the deployment script merely appends the desired tag to it. This also implies that tags change on each deployment, allowing automatic rollbacks to work.

Finally, my services initially point to the "template" task definition, but include:

```hcl
  lifecycle {
    ignore_changes = [task_definition, load_balancer]
  }
```

This means that after the first deploy, ECS replaces the task definition for the service, and terraform never touches that again. Ever.

I've been using this setup and found this setup works really well. Deployments never result in any noise in terraform plans, and terraform itself FULLY manages the template, while deployments operate on a separate TD.

**Comment by WhyNotHugo** [comment]
@anGie44 I understand that terraform won't make changes to support this odd-case. Do you mind **not** locking this issue, so workarounds can continue to be discussed in comments here (including, probably, by others in the same situation in future).

**Comment by Sytten** [comment]
@anGie44 Terraform could provide some help to allow for a better user experience still.
There is currently no good way to tell it: check if that resource exists or create it otherwise. Some people proposed work around but they are not perfect and can create loops.
Since this is a very common use case I would expect something to be done about it, if not by the official aws provider then by some community driven provider. A won't fix is not helpful...

Looking at what I am doing with GCP Cloud Run, I can quite easily use the provider to change the memory size without changing the image that is used since that is managed by the CICD (like ecs-deploy). It is just a matter of considering the task definition as a whole state (modified using revisions) without pinning it to one revision. So you can ignore the changes in say image and still detect changes in memory/cpu.

**Comment by sworisbreathing** [comment]
@angie44 thank you for clearly explaining the team's position on this issue. I agree, ECS doesn't do a great job of assisting the developer experience here, due to the fact that task definitions are immutable and versioned.

I'd had it in the back of my mind for a while now to roll my own provider while #11506 was being ignored, but other work took priority and I was able to get by with manual hacks. Since this basically confirms the PR won't be accepted, it seems the community is left with little choice but to use a custom provider in order to make ECS work properly.

**Comment by sworisbreathing** [comment]
also @anGie44:
>having one resource responsible for multiple AWS resources... would prove problematic to traditional practitioner experience with the rest of the Terraform provider ecosystem

Isn't that exactly what the `aws_security_group` resource does when you define the rules as part of the resource, though?

**Comment by WhyNotHugo** [comment]
> Since this basically confirms the PR won't be accepted, it seems the community is left with little choice but to use a custom provider in order to make ECS work properly.

The "template" task definition approach works, and doesn't violate any of the principles that terraform or ECS follow. The biggest downside is that listing task definitions yields more results, so if you interact manually with ECS a lot, then that might be annoying.

**Comment by schmod** [comment]
> Patches that suggest setting or manipulating the state difference of an ECS Task Definition’s’ revision would imply having one resource responsible for multiple AWS resources

I'm kind of surprised that you're viewing task revisions as distinct resources.  Surely there's precedent in the Terraform ecosystem for managing resources that maintain immutable version-histories?

**Comment by WhyNotHugo** [comment]
>  Surely there's precedent in the Terraform ecosystem for managing resources that maintain immutable version-histories?

Lambdas.

> I'm kind of surprised that you're viewing task revisions as distinct resources.

There's two "dimensions" in which I make changes to task definitions:

- During infra changes, via Terraform. Changes only the image_url and version.
- During deployments, via CodeDeploy. Changes anything _except_ those too.

Having this concept of "Task definition template"and "Task definition" means that Terraform owns one of these, and CodeDeploy owns the other.
